### PR TITLE
fix: multiple import statements for `tslib`

### PR DESCRIPTION
**Pull request by alan-agius4**
Closes: #587

## I'm submitting a...

```
[x] Bug Fix
[ ] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [ ] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [x] Tests for the changes have been added

## Description

Cherry pick tslib methods and avoid having multiple imports for tslib

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/lib/flatten/flatten.ts [hunk 1]
```diff
@@ -57,8 +57,18 @@ export async function flattenToFesm15(opts: FlattenOpts): Promise<string> {
 
 export async function flattenToFesm5(opts: FlattenOpts): Promise<string> {
   const destFile = path.resolve(opts.outDir, 'esm5', opts.flatModuleFile + '.js');
+  // a temp file is required because of the sourcemaps remapping
+  const destTempFile = path.resolve(opts.outDir, 'esm5-downlevel-temp', opts.flatModuleFile + '.js');
 
-  await downlevelWithTsc(opts.entryFile, destFile);
+  await downlevelWithTsc(opts.entryFile, destTempFile);
+  // the below is required in order to avoid multiple imports for tslib
+  await rollupBundleFile({
```
Comment (dherges) [inline src/lib/flatten/flatten.ts]: Hm, I'd prefer to do the down-levelling then in a rollup plugin

For example with bublé : https://github.com/rollup/rollup-plugin-buble

Or register the tsc downlevelling as a transform rollup plugin https://github.com/rollup/rollup/wiki/Plugins#creating-plugins:

```ts
rollup.plugins = [
   ...existinPlugins,
   {
      transform: (source, id ) => {
         const transpiled = ts.transpileModule(/* ... */);

         return {
           code: transpiled.outputText,
           map: transpiled.sourceMap
         };
      }
   }
]
```

Comment (alan-agius4) [reply src/lib/flatten/flatten.ts]: It's cleaner I agree

Comment (dherges) [reply src/lib/flatten/flatten.ts]: Should I update your PR?

Comment (alan-agius4) [reply src/lib/flatten/flatten.ts]: I updated my PR with some code changes as suggested. However, If you need to do some more changes. Feel free.

Comment (dherges) [reply src/lib/flatten/flatten.ts]: Oh, you already did!

File: src/lib/flatten/tsc.ts [hunk 2]
```diff
@@ -1,48 +1,29 @@
...
-export async function downlevelWithTsc(inputFile: string, outputFile: string): Promise<void> {
-  debug(`tsc ${inputFile} to ${outputFile}`);
-  const inputBuffer: Buffer = await readFile(inputFile);
-  const input: string = inputBuffer.toString();
+export async function downlevelWithTsc(code: string, filePath: string): Promise<SourceDescription> {
+  debug(`tsc ${filePath}`);
   const compilerOptions: CompilerOptions = {
     target: ScriptTarget.ES5,
     module: ModuleKind.ES2015,
     allowJs: true,
```
Comment (alan-agius4) [inline src/lib/flatten/tsc.ts]: I am not sure, why this is needed, in general, libraries that generate `.d.ts` cannot use this.

Comment (dherges) [reply src/lib/flatten/tsc.ts]: What do you mean "this"?

Comment (dherges) [reply src/lib/flatten/tsc.ts]: I am not sure what you refer to with "this". Is it the sourcemaps? I remember that was needed as it was transpiling the fesm15 file and w/o stripping it had two "// sourceMappingUrl" lines at end of fesm5 file.

Comment (alan-agius4) [reply src/lib/flatten/tsc.ts]: This was just as a side note. `allowJs` usually shouldn't be enabled. for libraries as libraries usually generate `d.ts` with `declaration: true` and this flag doesn't work with `declaration: true`.

So in theory I don't see the point of having `allowJs` in a library that generated `d.ts`

Comment (dherges) [reply src/lib/flatten/tsc.ts]: `allowJs` is for adding `.js` files to the compilation as we're down-levelling from a `*.js` file (not a `*.ts.`)

> Allow JavaScript files to be compiled.

https://www.typescriptlang.org/docs/handbook/compiler-options.html

Comment (alan-agius4) [reply src/lib/flatten/tsc.ts]: Ah thanks mate for the info

File: src/lib/flatten/tsc.ts [hunk 3]
```diff
@@ -1,48 +1,29 @@
...
-  ]);
+  return {
+    code: transpiled.outputText,
+    map: transpiled.sourceMapText
+  };
 }
-
-const REGEXP = /\/\/# sourceMappingURL=.*\.js\.map/;
-const trimSourceMap = (fileContent: string): string => {
-  if (fileContent.match(REGEXP)) {
```
Comment (dherges) [inline src/lib/flatten/tsc.ts]: Did you check the sourcemaps generated in the dist folder?  There have been issues in the past when an incorrect source map was going to fail the consumer's build.

I am going to double-check later that we're "safe" here

**Comment by alan-agius4** [comment]
They look good to me, but to be honest. I don't know what was the problem.
So, if you could check that would be great

File: src/lib/flatten/tsc.ts (hunk 3 above)
Comment (alan-agius4) [reply src/lib/flatten/tsc.ts]: They look good to me, but to be honest. I don't know what was the problem.
So, if you could check that would be great

Link to the artifacts;
https://1301-91553810-gh.circle-artifacts.com/0/integration/samples/core/dist/esm5/sample-core.js.map

Comment (alan-agius4) [reply src/lib/flatten/tsc.ts]: any update please?

Comment (dherges) [reply src/lib/flatten/tsc.ts]: I am looking at some source map files with https://sokra.github.io/source-map-visualization/#custom

the ones from the jsx sample look good!

Comment (dherges) [reply src/lib/flatten/tsc.ts]: material also. let' merge!

**Pull request merged**